include/dca/phys/dca_step/cluster_solver/ctint/structs/solver_configuration.hpp

```diff
@@ -0,0 +1,391 @@
+      doDoubleUpdate(rng)) {  // Double removal.
+    partners_lists_.clear();
+    for (const auto& partner_id : (*H_int_)[vertices_[candidates[0]].interaction_id].partners_id)
+      partners_lists_.push_back(&existing_[partner_id]);
+
+    const auto tag = details::getRandomElement(partners_lists_, rng());
+    candidates.push_back(findTag(tag));
+    assert(candidates[1] < int(size()) && candidates[1] >= 0);
+  }
+
+  assert(candidates[0] < int(size()));
+  return candidates;
+}
+
+void SolverConfiguration::push_back(const Vertex& v) {
+  vertices_.push_back(v);
+  BaseClass::addVertex(v);
+}
+
+void SolverConfiguration::pop(const int n) {
+  assert(n <= (int)size());
+
+  const int first_idx = size() - n;
+  std::array<int, 2> removal_size{0, 0};
+  for (std::size_t i = first_idx; i < size(); ++i)
+    addSectorSizes(i, removal_size);
+  vertices_.erase(vertices_.begin() + first_idx, vertices_.end());
+  BaseClass::pop(removal_size[0], removal_size[1]);
+}
+
+std::array<int, 2> SolverConfiguration::sizeIncrease() const {
+  assert(last_insertion_size_ > 0);
+  std::array<int, 2> result{0, 0};
+  for (std::size_t i = size() - last_insertion_size_; i < size(); ++i)
+    addSectorSizes(i, result);
+
+  return result;
+}
+
+inline void SolverConfiguration::addSectorSizes(int idx, std::array<int, 2>& sizes) const {
+  auto spin = [=](ushort nu) { return nu >= n_bands_; };
+  const auto& nu = coordinates(idx).nu;
+  ++sizes[spin(nu[0])];
+  ++sizes[spin(nu[2])];
+}
+
+inline short SolverConfiguration::getSign(const int vertex_index) const {
+  return getStrength(vertex_index) > 0 ? 1 : -1;
+}
+
+inline double SolverConfiguration::getStrength(int vertex_index) const {
+  assert(vertex_index >= 0 && vertex_index < (int)size());
+  return (*H_int_)[vertices_[vertex_index].interaction_id].w;
+}
+
+int SolverConfiguration::nPartners(int vertex_index) const {
+  assert(vertex_index < vertices_.size());
+  const auto& partners_id = (*H_int_)[vertices_[vertex_index].interaction_id].partners_id;
+  int n_partners = 0;
+  for (const auto partner_id : partners_id)
+    n_partners += existing_[partner_id].size();
+  return n_partners;
+}
+
+template <class Alloc>
+inline void SolverConfiguration::moveAndShrink(std::array<std::vector<int, Alloc>, 2>& sector_from,
+                                               std::array<std::vector<int, Alloc>, 2>& sector_to,
+                                               std::vector<int>& from, std::vector<int>& to) {
+  for (int s = 0; s < 2; ++s) {
+    // Sort and prepare source array.
+    auto& sector = BaseClass::sectors_[s].entries_;
+    std::sort(sector_to[s].begin(), sector_to[s].end());
+    auto& tags = BaseClass::sectors_[s].tags_;
+    sector_from[s].clear();
+    int source_idx = sector.size() - sector_to[s].size();
+    for (int i = 0; source_idx < sector.size(); ++i, ++source_idx) {
+      while (std::binary_search(sector_to[s].begin(), sector_to[s].end(), source_idx))
```

Comment:
I found these dead i indexes, while loops and checks for sector size overun very confusion. By the time I figured out they were correct (I think) i had figured out a significantly easier way to write them.  See following PR. I pass the test so if my PR is wrong then the testing is insufficient.

Reply:
Removed index i in 35b02bc5a9f92c7d3a74ad038a94b299ee0f3fe1